### PR TITLE
Enable multi-architecture docker image builds

### PR DESCRIPTION
**Pull request by tkw1536**
Previously, the Dockerfile downloaded 'docker-gen' and 'forego' binaries during build time. This caused a problem as it hard-coded the amd64 architecture for the images.

This PR updates both 'Dockerfile' and 'Dockerfile.alpine' to build the `forego` and `docker-gen` executables from source instead of downloading binaries directly. This is achieved using multi-stage builds [1]. Two separate stages first build the binaries, and are then copied over to the final stage.

The advantage of this change is two-fold: First, it enables building this image on architectures other than amd64. Second it adds trust by not adding external binaries to the docker image.

This modified version passes the tests both a linux desktop (`amd64`) as well as a raspberry pi (`armv7`) with some caveats:

- On `armv7`, a modified version of the `jwilder/docker-gen` image is required. See jwilder/docker-gen#327. 
- The `test_dhparam_is_generated_if_missing` test fails. This also doesn't currently pass on master. See #1468. 

See also:
- #1317 and #1370, #1297: These issues request an arm version of the image. Together with publishing arm images on DockerHub, this PR should fix all of them. 
- #1262, #1071: Both these PRs make a separate Dockerfile for the arm architecture. This PR is more general, in that it doesn't assume a specific architecture. In particular any architecture supported by the golang, debian and alpine base images can be used. 

[1] https://docs.docker.com/develop/develop-images/multistage-build/

### CODE REVIEW

**Comment by dustinrue** [comment]
I'd also really like to see this implemented

**Comment by awehrfritz** [comment]
Excellent initiative and nice implementation! @jwilder is there a chance to get this merged into master or do you have any objections to this?

**Comment by tkw1536** [comment]
I have yet to hear any feedback from @jwilder. I have nonetheless rebased this onto the current master, I only had to update the nginx version number.

**Comment by plonkus** [comment]
This PR helped me get up and running on the Apple M1 processor. Thank you @tkw1536

**Comment by kipusoep** [comment]
So any instructions to start using this on armv7?
I tried to build the docker image from your repo, which works, but still got errors running the image.
Then I saw your note about `jwilder/docker-gen`, how do I incorporate that fix in this fix as well?

**Comment by tkw1536** [comment]
@kipusoep This image should just build and run on armv7, it did when I ran the tests under qemu locally. What is the error you are getting?

**Comment by danil-smirnov** [comment]
I think this should be considered, very useful change

**Comment by danil-smirnov** [comment]
@tkw1536 I believe there is an issue with this branch.

Original @jwilder code sets `DOCKER_GEN_VERSION` environment variable, which `letsencrypt-nginx-proxy-companion` product relies on, see

https://github.com/nginx-proxy/docker-letsencrypt-nginx-proxy-companion/blob/master/app/functions.sh#L241

After your multistage build this environment variable appeared unset, so `letsencrypt-nginx-proxy-companion` container fails.

**Comment by tkw1536** [comment]
@danil-smirnov  I'm not convinced this is a problem. If you look at the diff of the PRs, neither of the old Dockerfiles set this variable and they seem to work fine on `linux/amd64` for many people.

**Comment by danil-smirnov** [comment]
@tkw1536 Okay, let's look at the old Dockerfiles:

https://github.com/nginx-proxy/nginx-proxy/blob/master/Dockerfile#L21
https://github.com/nginx-proxy/nginx-proxy/blob/master/Dockerfile.alpine#L18

Both set this variable so `letsencrypt-nginx-proxy-companion` can rely on them.

I've added the line `ENV DOCKER_GEN_VERSION 0.7.4` to my custom image and `letsencrypt-nginx-proxy-companion` can started working after the fix.

**Comment by tkw1536** [comment]
Oh you are correct. I forgot to read that `letsencrypt-nginx-proxy-companion` relies on this variable. 
I will add it to the branch.

**Comment by tkw1536** [comment]
Added it to the branch, thanks @danil-smirnov.
Not all we have to do is wait for this PR to be merged.

**Comment by severin-bruhat** [comment]
Hello, any update on this one?
Thanks

**Comment by awehrfritz** [comment]
> Now all we have to do is wait for this PR to be merged.

We are waiting already for quite a while... more than half a year. It's a bit worrying that the last commit in this repository is from Oct 2020. Not sure who controls this repository or who else other than @jwilder has is able to merge PRs.

**Comment by tkw1536** [comment]
Rebased onto the current master. With a new maintainer I would be happy to see this merged.

**Review by buchdag — CHANGES_REQUESTED** [review]
I'd rather use directly `FROM golang:1.15.x AS gobuilder` rather than passing the base image version used with `ARG`, so that we can use `dependabot`. 

Other than that, looks good to me.

**Comment by tkw1536** [comment]
I've updated this pull request with the changes requested.

**Comment by buchdag** [comment]
@tkw1536 could you do the same for the `FROM nginx:$NGINX_VERSION`, for the same reason (detection of dependencies versions by dependabot) ?

**Comment by tkw1536** [comment]
Done, updated. I don't think it's needed for `DOCKER_GEN_VERSION` or `FOREGO_VERSION`, as those are probably not covered by dependabot.

**Review by buchdag — APPROVED** [review]
LGTM

**Comment by buchdag** [comment]
@tkw1536 if you spotted other arm related issue or PR that might be closed following the merging of this PR, don't hesitate to tag me on them.

**Pull request merged**
